spring-boot-project/spring-boot-autoconfigure/src/main/java/org/springframework/boot/autoconfigure/flyway/NativeImageResourceProviderCustomizer.java

```diff
@@ -24,26 +26,74 @@
 import org.flywaydb.core.internal.scanner.ResourceNameCache;
 import org.flywaydb.core.internal.scanner.Scanner;
 
+import org.springframework.util.ClassUtils;
+
 /**
  * Registers {@link NativeImageResourceProvider} as a Flyway
  * {@link org.flywaydb.core.api.ResourceProvider}.
  *
  * @author Moritz Halbritter
+ * @author Maziz Esa
  */
 class NativeImageResourceProviderCustomizer extends ResourceProviderCustomizer {
 
+	public static final int NUM_FLYWAY10_SCANNER_PARAMETERS = 5;
+
 	@Override
 	public void customize(FluentConfiguration configuration) {
 		if (configuration.getResourceProvider() == null) {
-			Scanner<JavaMigration> scanner = new Scanner<>(JavaMigration.class,
-					Arrays.asList(configuration.getLocations()), configuration.getClassLoader(),
-					configuration.getEncoding(), configuration.isDetectEncoding(), false, new ResourceNameCache(),
-					new LocationScannerCache(), configuration.isFailOnMissingLocations());
+			Constructor<?> scannerConstructor;
+			Scanner scanner;
+			try {
+				scannerConstructor = ClassUtils.forName("org.flywaydb.core.internal.scanner.Scanner", null)
```

Comment:
I think it would be better to avoid reflection in the default case when using Flyway 9.

One way to do that would be to catch the `NoSuchMethodError` and then fall back to reflection. That fallback could even just assume that it's Flyway 10 and assume that `Scanner` will take 5 constructor arguments.